File: Xamarin/LightOrganApp.iOS/FileListViewController.cs
using Foundation;
using MediaPlayer;
using System;
using System.Collections.Generic;
using UIKit;
using System.Linq;
using System.Threading.Tasks;

namespace LightOrganApp.iOS
{
    public partial class FileListViewController : UITableViewController
    {
        enum RestorationKeys
        {
            SearchControllerIsActive,
            SearchBarText,
            SearchBarIsFirstResponder
        }

        struct SearchControllerRestorableState
        {
            public bool WasActive { get; set; }
            public bool WasFirstResponder { get; set; }
        }

        static NSString cellId = new NSString("reuseIdentifier");

        List<MPMediaItem> allMediaItems;
        List<MPMediaItem> filteredMediaItems;
        List<MPMediaItem> selectedMediaItems;
        public MPMediaItemCollection didPickMediaItems;

        UISearchController searchController;

        SearchControllerRestorableState restoredState = new SearchControllerRestorableState();

        public FileListViewController (IntPtr handle) : base (handle)
        {
        }

        public async override void ViewDidLoad()
        {
            base.ViewDidLoad();

            ConfigureSearchController();            

            TableView.Source = new TableSource(this);

            TableView.TableFooterView = new UIView();
            TableView.BackgroundView = new UIView();

            selectedMediaItems = new List<MPMediaItem>();

            if (UIDevice.CurrentDevice.CheckSystemVersion(9, 3))
            {
                var status = await MPMediaLibrary.RequestAuthorizationAsync();

                if (status == MPMediaLibraryAuthorizationStatus.Authorized)
                    LoadMediaItemsForMediaTypeAsync(MPMediaType.Music);
                else
                    DisplayMediaLibraryError();                
            }
            else
            {
                LoadMediaItemsForMediaTypeAsync(MPMediaType.Music);
            }      
        }

        public override void ViewDidAppear(bool animated)
        {
            base.ViewDidAppear(animated);

            if (restoredState.WasActive)
            {
                searchController.Active = restoredState.WasActive;
                restoredState.WasActive = false;

                if (restoredState.WasFirstResponder)
                {
                    searchController.SearchBar.BecomeFirstResponder();
                    restoredState.WasFirstResponder = false;
                }
            }
        }

        [Export("searchBarSearchButtonClicked:")]
        public virtual void SearchButtonClicked(UISearchBar searchBar)
        {
            searchBar.ResignFirstResponder();
        }

        private void ConfigureSearchController()
        {
            searchController = new CustomSearchController((UIViewController)null)
            {
                WeakSearchResultsUpdater = this
            };
            searchController.DimsBackgroundDuringPresentation = false;
            searchController.SearchBar.SizeToFit();
            UISearchBar.Appearance.TintColor = UIColor.FromRGB(197, 225, 165);
            searchController.SearchBar.Placeholder = NSBundle.MainBundle.LocalizedString("searchMusic", "Search Music");
            searchController.SearchBar.WeakDelegate = this;
            DefinesPresentationContext = true;
            NavigationItem.TitleView = searchController.SearchBar;
            searchController.HidesNavigationBarDuringPresentation = false;
        }

        private async void LoadMediaItemsForMediaTypeAsync(MPMediaType mediaType)
        {
            await Task.Run(() =>
            {
                var query = new MPMediaQuery();
                var mediaTypeNumber = NSNumber.FromInt32((int)mediaType);
                var predicate = MPMediaPropertyPredicate.PredicateWithValue(mediaTypeNumber, MPMediaItem.MediaTypeProperty);

                query.AddFilterPredicate(predicate);

                allMediaItems = query.Items.ToList();
            });

            TableView.ReloadData();                       
        }

        private List<MPMediaItem> GetMediaItems()
        {
            if ((searchController.Active || restoredState.WasActive) && searchController.SearchBar.Text != "")
                return filteredMediaItems;
            else           
                return allMediaItems;
        }

        private void DisplayMediaLibraryError()
        {
            string error;

            switch (MPMediaLibrary.AuthorizationStatus)
            {
                case MPMediaLibraryAuthorizationStatus.Restricted:
                    error = "Media library access restricted by corporate or parental settings";
                    break;
                case MPMediaLibraryAuthorizationStatus.Denied:
                    error = "Media library access denied by user";
                    break;
                default:
                    error = "Unknown error";
                    break;
            }

            var controller = UIAlertController.Create("Error", error, UIAlertControllerStyle.Alert);
            controller.AddAction(UIAlertAction.Create("OK", UIAlertActionStyle.Default, null));

            PresentViewController(controller, true, null);               
        }

        public override void PrepareForSegue(UIStoryboardSegue segue, NSObject sender)
        {
            if (sender == doneButton)
            {
                if (selectedMediaItems != null && selectedMediaItems.Count > 0)
                    didPickMediaItems = new MPMediaItemCollection(selectedMediaItems.ToArray());                
            }
        }

        private bool MediaItemContainsString(MPMediaItem item, string searchText)
        {
            var b1 = false;
            var title = item.Title;
            if (title != null)
                b1 = title.ToString().ToLower().Contains(searchText.ToLower());

            var b2 = false;
            var artist = item.Artist;
            if (artist != null)
                b2 = artist.ToString().ToLower().Contains(searchText.ToLower());

            return b1 || b2;
        }

        private void FilterContentForSearchText(string searchText)
        {
            if (allMediaItems == null)
                return;

            filteredMediaItems = allMediaItems.Where(item => MediaItemContainsString(item, searchText)).ToList();

            TableView.ReloadData();
        }

        [Export("updateSearchResultsForSearchController:")]
        public virtual void UpdateSearchResultsForSearchController(UISearchController searchController)
        {
            FilterContentForSearchText(searchController.SearchBar.Text);
        }

        public override void EncodeRestorableState(NSCoder coder)
        {
            base.EncodeRestorableState(coder);

            coder.Encode(searchController.Active, RestorationKeys.SearchControllerIsActive.ToString());
            coder.Encode(searchController.SearchBar.IsFirstResponder, RestorationKeys.SearchBarIsFirstResponder.ToString());
            coder.Encode(new NSString(searchController.SearchBar.Text), RestorationKeys.SearchBarText.ToString());
        }

        public override void DecodeRestorableState(NSCoder coder)
        {
            base.DecodeRestorableState(coder);

            restoredState.WasActive = coder.DecodeBool(RestorationKeys.SearchControllerIsActive.ToString());
            restoredState.WasFirstResponder = coder.DecodeBool(RestorationKeys.SearchBarIsFirstResponder.ToString());
            searchController.SearchBar.Text = (NSString)coder.DecodeObject(RestorationKeys.SearchBarText.ToString());
        }

        class TableSource : UITableViewSource
        {
            FileListViewController controller;

            public TableSource(FileListViewController controller)
            {
                this.controller = controller;
            }

            public override nint RowsInSection(UITableView tableView, nint section)
            {
                var mediaItems = controller.GetMediaItems();

                if (mediaItems != null)
                    return mediaItems.Count;
                else
                    return 0;                
            }

            public override void WillDisplay(UITableView tableView, UITableViewCell cell, NSIndexPath indexPath)
            {
                cell.TintColor = UIColor.White;

                if (cell.TextLabel != null)
                    cell.TextLabel.TextColor = UIColor.White;

                if (cell.DetailTextLabel != null)
                    cell.DetailTextLabel.TextColor = UIColor.FromRGB(197, 225, 165);
            }

            public override UITableViewCell GetCell(UITableView tableView, NSIndexPath indexPath)
            {
                var cell = tableView.DequeueReusableCell(cellId, indexPath);
                var row = indexPath.Row;
                var mediaItems = controller.GetMediaItems();

                var item = mediaItems[row];
                if (cell.TextLabel != null)
                    cell.TextLabel.Text = item.Title;

                var artist = NSBundle.MainBundle.LocalizedString("unknownArtist", "Unknown Artist");
                var artistVal = item.Artist;
                if (artistVal != null)
                    artist = artistVal;

                var length = (int)item.PlaybackDuration;

                if (cell.DetailTextLabel != null)
                    cell.DetailTextLabel.Text = $"{artist} {GetDisplayTime(length)}";

                if (controller.selectedMediaItems != null && controller.selectedMediaItems.Contains(item))
                    cell.Accessory = UITableViewCellAccessory.Checkmark;
                else
                    cell.Accessory = UITableViewCellAccessory.None;

                cell.Tag = row;

                return cell;
            }

            private string GetDisplayTime(int seconds)
            {
                var h = seconds / 3600;
                var m = seconds / 60 - h * 60;
                var s = seconds - h * 3600 - m * 60;

                var str = "";

                if (h > 0)
                    str += $"{h}:";

                str += $"{m:00}:{s:00}";

                return str;
            }

            public override void RowSelected(UITableView tableView, NSIndexPath indexPath)
            {
                var row = indexPath.Row;
                var mediaItems = controller.GetMediaItems();
                var item = mediaItems[row];

                if (controller.selectedMediaItems == null || !controller.selectedMediaItems.Contains(item))                
                    controller.selectedMediaItems?.Add(item);                
                else                
                    controller.selectedMediaItems.Remove(item);              

                tableView.ReloadData();
            }
        }        
    }
}

File: iOS10/LightOrganApp/FileListViewController.swift
//
//  FileListViewController.swift
//  LightOrganApp
//
//  Created by Marcin Kruszyński on 03/05/16.
//  Copyright © 2016 Marcin Kruszyński. All rights reserved.
//

import UIKit
import MediaPlayer

class FileListViewController: UITableViewController, UISearchBarDelegate, UISearchResultsUpdating {
    
    enum RestorationKeys : String {
        case searchControllerIsActive
        case searchBarText
        case searchBarIsFirstResponder
    }
    
    struct SearchControllerRestorableState {
        var wasActive = false
        var wasFirstResponder = false
    }
    
    
    @IBOutlet var doneButton: UIBarButtonItem!
    
    var allMediaItems: [MPMediaItem]?
    var filteredMediaItems: [MPMediaItem]?
    var selectedMediaItems: [MPMediaItem]?
    var didPickMediaItems: MPMediaItemCollection?
    
    var searchController: UISearchController!
    
    var restoredState = SearchControllerRestorableState()
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.configureSearchController()
        self.tableView.tableFooterView = UIView()
        self.tableView.backgroundView = UIView()
        
        selectedMediaItems = [MPMediaItem]()
        
        if #available(iOS 9.3, *) {
            MPMediaLibrary.requestAuthorization { (status) in
                if status == .authorized {
                    self.loadMediaItemsForMediaType(.music)
                } else {
                    self.displayMediaLibraryError()
                }
            }
        }
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        if restoredState.wasActive {
            searchController.isActive = restoredState.wasActive
            restoredState.wasActive = false
            
            if restoredState.wasFirstResponder {
                searchController.searchBar.becomeFirstResponder()
                restoredState.wasFirstResponder = false
            }
        }
    }
    
    func searchBarSearchButtonClicked(_ searchBar: UISearchBar) {
        searchBar.resignFirstResponder()
    }
    
    func configureSearchController() {
        searchController = CustomSearchController(searchResultsController: nil)
        searchController.searchResultsUpdater = self
        searchController.dimsBackgroundDuringPresentation = false
        searchController.searchBar.sizeToFit()
        searchController.searchBar.barTintColor = .black
        
        searchController.searchBar.placeholder = NSLocalizedString("searchMusic", comment: "Search Music")
        searchController.searchBar.delegate = self
        definesPresentationContext = true
        navigationItem.titleView = searchController.searchBar
        searchController.hidesNavigationBarDuringPresentation = false;
    }
    
    func loadMediaItemsForMediaType(_ mediaType: MPMediaType) {
        
        let queue = DispatchQueue.global(qos: .default)
        
        queue.async {
            let query = MPMediaQuery()
            let mediaTypeNumber =  Int(mediaType.rawValue)
            let predicate = MPMediaPropertyPredicate(value: mediaTypeNumber,
                                                     forProperty: MPMediaItemPropertyMediaType)
            query.addFilterPredicate(predicate)
            
            self.allMediaItems = query.items
            
            DispatchQueue.main.async {
                self.tableView.reloadData()
            }
        }
    }
    
    fileprivate func getMediaItems() -> [MPMediaItem]? {
        if (searchController.isActive || restoredState.wasActive) && searchController.searchBar.text != "" {
            return filteredMediaItems
        } else {
            return allMediaItems
        }
    }
    
    @available(iOS 9.3, *)
    func displayMediaLibraryError() {
        var error: String
        switch MPMediaLibrary.authorizationStatus() {
        case .restricted:
            error = "Media library access restricted by corporate or parental settings"
        case .denied:
            error = "Media library access denied by user"
        default:
            error = "Unknown error"
        }
        
        let controller = UIAlertController(title: "Error", message: error, preferredStyle: .alert)
        controller.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
        present(controller, animated: true, completion: nil)
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let mediaItems = getMediaItems()
        if mediaItems != nil {
            return mediaItems!.count;
        } else {
            return 0
        }
    }
    
    override func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        cell.textLabel?.textColor = .white
        cell.detailTextLabel?.textColor = .lightGray
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "reuseIdentifier", for: indexPath)
        let row = (indexPath as NSIndexPath).row
        let mediaItems = getMediaItems()
        
        
        let item = mediaItems![row] as MPMediaItem
        cell.textLabel?.text = item.value(forProperty: MPMediaItemPropertyTitle) as! String?
        
        var artist = NSLocalizedString("unknownArtist", comment: "Unknown Artist")
        if let artistVal = item.value(forProperty: MPMediaItemPropertyArtist) as? String {
            artist = artistVal
        }
        
        let length = item.value(forProperty: MPMediaItemPropertyPlaybackDuration) as! Int
        
        
        cell.detailTextLabel?.text = "\(artist)  \(getDisplayTime(length))"
        
        if selectedMediaItems!.contains(item) {
            cell.accessoryType = .checkmark
        } else {
            cell.accessoryType = .none
        }
        
        
        cell.tag = row
        
        return cell
    }
    
    fileprivate func getDisplayTime(_ seconds: Int) -> String {
        
        let h = seconds / 3600
        let m = seconds / 60 - h * 60
        let s = seconds - h * 3600 - m * 60
        
        var str = "";
        
        if h > 0 {
            str += "\(h):"
        }
        str += String(format: "%02d:%02d", m, s)
        
        return str
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        
        let row = (indexPath as NSIndexPath).row
        let mediaItems = getMediaItems()
        let item = mediaItems![row] as MPMediaItem
        
        if !selectedMediaItems!.contains(item) {
            selectedMediaItems!.append(item)
        } else {
            if let index = selectedMediaItems!.index(of: item) {
                selectedMediaItems!.remove(at: index)
            }
        }
        
        tableView.reloadData()
    }
    
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        
        if doneButton === sender as? UIBarButtonItem {
            if selectedMediaItems!.count > 0 {
                self.didPickMediaItems = MPMediaItemCollection(items: selectedMediaItems!)
            }
        }
    }
    
    
    func mediaItemContainsString(_ item: MPMediaItem, searchText: String) -> Bool {
        var b1 = false
        if let title = item.value(forProperty: MPMediaItemPropertyTitle) as? String {
            b1 = title.lowercased().contains(searchText.lowercased())
        }
        
        var b2 = false
        if let artist = item.value(forProperty: MPMediaItemPropertyArtist) as? String {
            b2 = artist.lowercased().contains(searchText.lowercased())
        }
        
        return b1 || b2
    }
    
    
    func filterContentForSearchText(_ searchText: String) {
        if self.allMediaItems == nil {
            return
        }
        
        self.filteredMediaItems = self.allMediaItems!.filter { item in
            return mediaItemContainsString(item, searchText: searchText)
        }
        
        tableView.reloadData()
    }
    
    func updateSearchResults(for searchController: UISearchController) {
        filterContentForSearchText(searchController.searchBar.text!)
    }
    
    
    override func encodeRestorableState(with coder: NSCoder) {
        super.encodeRestorableState(with: coder)
        
        
        coder.encode(searchController.isActive, forKey:RestorationKeys.searchControllerIsActive.rawValue)
        
        coder.encode(searchController.searchBar.isFirstResponder, forKey:RestorationKeys.searchBarIsFirstResponder.rawValue)
        
        coder.encode(searchController.searchBar.text, forKey:RestorationKeys.searchBarText.rawValue)
    }
    
    override func decodeRestorableState(with coder: NSCoder) {
        super.decodeRestorableState(with: coder)        
        
        
        restoredState.wasActive = coder.decodeBool(forKey: RestorationKeys.searchControllerIsActive.rawValue)
        
        restoredState.wasFirstResponder = coder.decodeBool(forKey: RestorationKeys.searchBarIsFirstResponder.rawValue)
        
        searchController.searchBar.text = coder.decodeObject(forKey: RestorationKeys.searchBarText.rawValue) as? String
    }
}





